From a6c28fea9fa791a85b1809f00cc43d16b87f2b77 Mon Sep 17 00:00:00 2001
From: Alberto Mardegan <alberto.mardegan@canonical.com>
Date: Fri, 18 Dec 2015 18:16:14 +0300
Subject: [PATCH] Fix endianness issues

We cannot safely load a 32 bit integer into a struct field which expects
a 64 bit value: we must load it into a 32 bit integer first, and then
assign the value to the structure field.

This bug was causing test failures on s390x:
https://gitlab.com/accounts-sso/libaccounts-glib/issues/2
---
 libaccounts-glib/ag-account.c | 6 ++++--
 libaccounts-glib/ag-manager.c | 8 +++++---
 2 files changed, 9 insertions(+), 5 deletions(-)

diff --git a/libaccounts-glib/ag-account.c b/libaccounts-glib/ag-account.c
index 165ae25..478ed9e 100644
--- a/libaccounts-glib/ag-account.c
+++ b/libaccounts-glib/ag-account.c
@@ -291,8 +291,10 @@ _ag_account_build_dbus_changes (AgAccount *account, AgAccountChanges *changes,
 
     if (ts)
     {
-        g_variant_builder_add (&builder, "u", ts->tv_sec);
-        g_variant_builder_add (&builder, "u", ts->tv_nsec);
+        guint32 sec = ts->tv_sec;
+        guint32 nsec = ts->tv_nsec;
+        g_variant_builder_add (&builder, "u", sec);
+        g_variant_builder_add (&builder, "u", nsec);
     }
     g_variant_builder_add (&builder, "u", account->id);
     g_variant_builder_add (&builder, "b", changes->created);
diff --git a/libaccounts-glib/ag-manager.c b/libaccounts-glib/ag-manager.c
index 8e6a891..7c8673b 100644
--- a/libaccounts-glib/ag-manager.c
+++ b/libaccounts-glib/ag-manager.c
@@ -699,16 +699,18 @@ dbus_filter_callback (G_GNUC_UNUSED GDBusConnection *dbus_conn,
     if (!object_path_is_interesting (object_path, priv->object_paths))
         return;
 
-    memset (&ts, 0, sizeof (struct timespec));
+    guint32 sec, nsec;
     g_variant_get (msg,
                    "(uuubb&s@*)",
-                   &ts.tv_sec,
-                   &ts.tv_nsec,
+                   &sec,
+                   &nsec,
                    &account_id,
                    &created,
                    &deleted,
                    &provider_name,
                    &v_services);
+    ts.tv_sec = sec;
+    ts.tv_nsec = nsec;
 
     DEBUG_INFO ("path = %s, time = %lu-%lu (%p)",
                 object_path, ts.tv_sec, ts.tv_nsec,
-- 
2.6.2

